docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,222 @@
+
+1. Signal emission and handling is not very well checked.  At a language level, there is no way for widgets to describe what type of signals they can emit.  Similarly, there is no syntax for distinguishing between widget handlers and "normal" handlers, meaning that (in principal) a handler that is not intended to be a signal handler could end up getting called in a signal context if a child widget happens to start emitting a signal with a matching name.  Finally, there is no facility for checking that signal prototypes match signal handler prototypes.
+
+2. Similar issues to 1. above apply for event handling.
+
+3. In LCS, event handling propagates upward: if a button doesn't have an `MouseUp` handler, then the event gets passed on to a parent card, and so on.  Furthermore, there is the idea of explicit propagation: it is possible to do something in an event handler and then, for example, `pass MouseUp` to invoke the normal event propagation logic.
+
+   In LCB, event propagation is *not* currently automatic.  If a widget does not explicitly handle and re-`post` each possible event, then its script (or parent widget) never sees the event.
+
+4. Sometimes, it's important to be able to be able to prevent child widgets from seeing an event, or performing some pre-processing / post-processing around the normal event handling process.  The current LCB event handling scheme does not permit this.
+
+## Assumptions
+
+For the purposes of this document we assume that:
+
+* All connection of signal and event handlers is implicit, rather than explicit.
+
+* An instance of a widget as an engine control can be considered as a composed widget, comprising a script and the "real" widget instance (this makes sense when you consider the fact that the widget control exists, has a script, and can be manipulated even if the "real" widget was unable to be loaded for some reason).  This will be referred to as a "host widget" in the rest of this document.
+
+## Signals
+
+### Signal emission
+
+A widget's LCB source code must explicitly name, and provide prototypes for, the signals that it can emit [1]:
+
+    public signal <name:id>(<params:formals>)
+
+Signal declarations must be placed in module scope.  The signal **name** must be unique in the module.
+
+Signals are explicitly emitted by a widget using an `emit` statement:
+
+    emit <name:id>(<params:expressions>)
+
+It is a run-time error to emit a signal with the wrong number or types of emission parameters [2].  The compiler may statically verify that signals are emitted with compatible parameter lists.
+
+The `emit` statement may return before all the signal handlers for the signal have finished running [3].
+
+The signal handlers for a signal get run exactly once for each emission.  All of the handlers for an emission must complete before any of the handlers for a subsequent emission begin [4].
+
+The signal handlers for a signal may be run in any order.
+
```

Comment:
@trevordevore: Okay - so your use-case is one where all the data is in memory, but needs processing to turn it into a form for the view to display; thus you use a callback to format data rather than to fetch.

I do wonder whether the callback approach does just 'shift' the 'smooth scrolling' problem down the chain (to the scripter) as it is then entirely a function of the speed of the format request for each row. So, in a case where there is a significant amount of processing required to generate the formatted data from the cached data the application holds (which has previously been fetched from the database) the application will have to hold cached formatted data in addition to the original data - the cost to format this data has to be paid at some point, most likely as the pagination occurs. (Although I appreciate that is handled in the current model as it currently is prepared for a pause at page breaks).

So, to my mind, one way to potentially go about this would be to have the view cache the formatted data (and precache, as necessary) and just display holding rows if it hasn't yet been populated. This means that the formatted data requests can be asynchronous and so it doesn't matter what the requests have to do to get the data to the view. This then means you can then have a view either visibly paginates (pauses at visible points and then extends), or one which scrolls completely smoothly but with the usual 'holding' image turning into real data a small delay after (which is quite a common thing you see in modern apps).

Now, of course, you have a great deal more experience writing dataview type controls than me - so I'm more than willing to be 'shot down in flames' :)

I'm sure it is possible to allow 'synchronous' signals safely - however, they will always come with limitations in what can be done within them. Basically, if a synchronous signal does anything which causes the event loop to be touched chances are it might trigger a redraw and as such using the control becomes 'slightly' more fragile.